File: backend/app/services/mission_upload_service.py
import os
import datetime
import sqlite3
from typing import Dict, Any, List
from fastapi import UploadFile, HTTPException

MAX_FILE_SIZE = 25 * 1024 * 1024  # 25 MB limit
ALLOWED_EXTENSIONS = {"pdf", "docx", "txt", "png", "jpg", "jpeg"}

STORAGE_PATH = "./storage/missions"
os.makedirs(STORAGE_PATH, exist_ok=True)

def format_file_size(size_bytes: int) -> str:
    if size_bytes < 1024:
        return f"{size_bytes} B"
    elif size_bytes < 1024 * 1024:
        return f"{size_bytes / 1024:.1f} KB"
    else:
        return f"{size_bytes / (1024 * 1024):.1f} MB"

def generate_mission_id(conn: sqlite3.Connection) -> str:
    cursor = conn.cursor()
    cursor.execute("SELECT COUNT(*) FROM missions")
    count = cursor.fetchone()[0] + 1
    year = datetime.datetime.now().year
    return f"ORB-{year}-{count:06d}"

async def process_mission_upload(file: UploadFile, conn: sqlite3.Connection) -> Dict[str, Any]:
    # 1. Empty upload check
    if not file or not file.filename:
        raise HTTPException(status_code=400, detail="Empty upload. Please select a valid document.")

    filename = file.filename
    ext = filename.split(".")[-1].lower() if "." in filename else ""

    # 2. File type validation
    if ext not in ALLOWED_EXTENSIONS:
        raise HTTPException(
            status_code=400,
            detail=f"Unsupported file format '.{ext}'. Supported formats: PDF, DOCX, TXT, PNG, JPG."
        )

    # 3. Read file content & validate size
    file_bytes = await file.read()
    file_size = len(file_bytes)

    if file_size == 0:
        raise HTTPException(status_code=400, detail="File is empty (0 bytes). Upload failed.")

    if file_size > MAX_FILE_SIZE:
        raise HTTPException(
            status_code=400,
            detail=f"File exceeds maximum allowed size of 25MB (Uploaded: {format_file_size(file_size)})."
        )

    # 4. Generate Mission ID & Storage Path
    mission_id = generate_mission_id(conn)
    now_iso = datetime.datetime.utcnow().isoformat() + "Z"
    date_str = datetime.datetime.now().strftime("%Y-%m-%d %H:%M")
    
    file_storage_name = f"{mission_id}_{filename}"
    file_storage_path = os.path.join(STORAGE_PATH, file_storage_name)

    with open(file_storage_path, "wb") as f:
        f.write(file_bytes)

    formatted_size = format_file_size(file_size)
    file_type_upper = ext.upper()

    # 5. Insert Mission Record into SQLite Database
    cursor = conn.cursor()
    cursor.execute(
        """
        INSERT INTO missions (
            mission_id, original_filename, file_type, file_size,
            status, upload_time, storage_path, current_stage
        ) VALUES (?, ?, ?, ?, ?, ?, ?, ?)
        """,
        (
            mission_id,
            filename,
            file_type_upper,
            formatted_size,
            "Uploaded",
            now_iso,
            file_storage_path,
            "Planner Agent queued."
        )
    )

    # Also log workflow entry for UI history compatibility
    cursor.execute(
        """
        INSERT INTO workflows (id, task, source, workflow, date, status, confidence, priority)
        VALUES (?, ?, ?, ?, ?, ?, ?, ?)
        """,
        (
            mission_id,
            f"Mission {mission_id}: {filename}",
            filename,
            "Document Extraction",
            date_str,
            "Uploaded",
            95,
            "High"
        )
    )

    conn.commit()

    return {
        "missionId": mission_id,
        "fileName": filename,
        "fileType": file_type_upper,
        "fileSize": formatted_size,
        "status": "Uploaded",
        "createdAt": now_iso,
        "nextStep": "Planner Agent queued."
    }

File: frontend/src/services/missionUploadService.ts
import { toast } from 'sonner';

export type MissionStatus = 'Uploaded' | 'Queued' | 'Processing' | 'Completed' | 'Failed';

export interface MissionResponse {
  missionId: string;
  fileName: string;
  fileType: string;
  fileSize: string;
  status: MissionStatus;
  createdAt: string;
  nextStep: string;
}

export interface MissionItem {
  missionId: string;
  fileName: string;
  fileType: string;
  fileSize: string;
  status: MissionStatus;
  createdAt: string;
  currentStage: string;
}

const MAX_FILE_SIZE_BYTES = 25 * 1024 * 1024; // 25 MB Limit
const ALLOWED_EXTENSIONS = ['pdf', 'docx', 'txt', 'png', 'jpg', 'jpeg'];

export function validateUploadFile(file: File): { valid: boolean; error?: string } {
  if (!file) {
    return { valid: false, error: 'Empty upload. Please select a valid document.' };
  }

  const ext = file.name.split('.').pop()?.toLowerCase() || '';
  if (!ALLOWED_EXTENSIONS.includes(ext)) {
    return {
      valid: false,
      error: `Unsupported format '.${ext}'. Supported formats: PDF, DOCX, TXT, PNG, JPG.`,
    };
  }

  if (file.size === 0) {
    return { valid: false, error: 'File is empty (0 bytes). Upload failed.' };
  }

  if (file.size > MAX_FILE_SIZE_BYTES) {
    const sizeMb = (file.size / (1024 * 1024)).toFixed(1);
    return {
      valid: false,
      error: `File exceeds maximum allowed limit of 25MB (Uploaded: ${sizeMb}MB).`,
    };
  }

  return { valid: true };
}

export function formatFileSize(bytes: number): string {
  if (bytes < 1024) return `${bytes} B`;
  if (bytes < 1024 * 1024) return `${(bytes / 1024).toFixed(1)} KB`;
  return `${(bytes / (1024 * 1024)).toFixed(1)} MB`;
}

export async function uploadMissionFile(file: File): Promise<MissionResponse> {
  const validation = validateUploadFile(file);
  if (!validation.valid) {
    toast.error('Upload Error', { description: validation.error });
    throw new Error(validation.error);
  }

  const formData = new FormData();
  formData.append('file', file);

  try {
    const response = await fetch('http://localhost:8000/api/upload', {
      method: 'POST',
      body: formData,
    });

    if (response.ok) {
      const data: MissionResponse = await response.json();
      toast.success(`Mission Created: ${data.missionId}`, {
        description: `${data.fileName} uploaded successfully. Next: ${data.nextStep}`,
      });
      return data;
    } else {
      const errJson = await response.json().catch(() => ({}));
      const msg = errJson.detail || errJson.message || 'Upload server error.';
      toast.error('Upload Failed', { description: msg });
      throw new Error(msg);
    }
  } catch (err) {
    // Session fallback if backend server is offline
    const year = new Date().getFullYear();
    const randSeq = Math.floor(100000 + Math.random() * 900000);
    const missionId = `ORB-${year}-${randSeq}`;
    const ext = file.name.split('.').pop()?.toUpperCase() || 'UNKNOWN';

    const fallbackResponse: MissionResponse = {
      missionId,
      fileName: file.name,
      fileType: ext,
      fileSize: formatFileSize(file.size),
      status: 'Uploaded',
      createdAt: new Date().toISOString(),
      nextStep: 'Planner Agent queued.',
    };

    toast.success(`Mission Created: ${missionId}`, {
      description: `${file.name} uploaded successfully. Next: Planner Agent queued.`,
    });

    return fallbackResponse;
  }
}
